vt_graph_api/graph.py

```diff
@@ -1750,12 +1869,88 @@ def download_screenshot(self, path = "."):
           )
       )
 
+
   def _get_nodes_by_type(self, node_type):
     return {"data": [
         {'type': node.node_type, 'id': node.node_id}
         for node in self.nodes.values() if node.node_type == node_type
     ]}
 
+
+  def _get_groups_nodes(self, final_nodes):
+    """Returns all the loaded and generated group nodes.
+
+    Args:
+      final_nodes: Set with the ids of all the added nodes.
+
+    Returns: Array of group nodes.
+
+    """
+    group_nodes = []
+    current_index = len(final_nodes)
+    for idx, group_node in enumerate(self.group_nodes.values()):
+      belonging_nodes = group_node['entity_attributes']['grouped_node_ids']
+      belonging_nodes = list(
+          set([node for node in belonging_nodes if node in final_nodes]))
+      if (belonging_nodes):
+        group_node['entity_attributes']['grouped_node_ids'] = belonging_nodes
+        group_node['index'] = current_index + idx
+        group_nodes.append(group_node)
+    return group_nodes
+
+
+  def _get_groups_links(self, final_links):
+    """Generates all the links needed to represent groups in VTGraph.
+
+    Args:
+      final_links: list with all the links added in the Graph.
+
+    Returns: List with the links needed to represent groups in VTGraph.
+
+    """
+    map_node_group = {}
+
+    # Generate a link for each pair node_in_group -> group_node
+    for group_node in self.group_nodes.values():
+      for node_in_group in group_node['entity_attributes']['grouped_node_ids']:
+        map_node_group[node_in_group] = group_node['entity_id']
+
+    group_links = [{'source': node, 'target': group, 'connection_type': 'group'}
+                   for node, group in map_node_group.items()]
+
+    # Generate a link for each link where source or target is a node
+    # inside a group.
+    for link in final_links:
+      if link['source'] in map_node_group:
+        new_link = link.copy()
+        new_link['source'] = map_node_group[link['source']]
+        group_links.append(new_link)
+      elif link['target'] in map_node_group:
+        new_link = link.copy()
+        new_link['target'] = map_node_group[link['target']]
+        group_links.append(new_link)
+    return group_links
+
+
+  def _get_special_relationship_nodes(self):
+    """Returns a list with the special relationship nodes."""
+    return list(self.special_relationship_nodes.values())
+
+
+  def _get_special_relationship_links(self, final_nodes):
```

Comment:
_get_special_relationship_nodes retrieves nodes from self. In the case of links we are receiving a parameter. Are they from the same flow?